### PR TITLE
Exclude Properties Using Exclusion List #113

### PR DESCRIPTION
**Pull request by deev-pal08**
# Exclude Properties from URI Using Exclusion List

Exclusion List is a feature which allows you to exclude/restrict certain entities and their respective properties while generating URL payloads for entity_sets parsed by pyodata from the metadata.xml file

This feature has been implemented in the DirectBuider class of the odfuzz/entities.py file.

Earlier when we mentioned a particular property in the ExclusionList, it was only getting restricted from the Body of the generated Payload, but it should also be restricted from the URI of the generated payloads.

While using this Exclusion_List we can mention certain properties for a particular EntitySet which we do not want while generating the payloads.

**Example**:

 ```json
              {
                "$ENTITY_SET$":{
                        "Products": {
                            "Properties":["ProductID", "ProductName"],
                            "Nav_Properties":["Category"]
                        }
                    }
                }
```

Two functions use these properties to generate the payloads and include them in the URI:
1. **$filter**
2. **$orderby**

Implemented regex_match function in two of the classes which uses $orderby and $filter functions to generate the payloads for the URI:

1. **FilterQuery**: Class to generate queries for $filter function
2. **OrderbyQuery**: Class to generate queries for $orderby function

Both the classes randomly choose properties from the list of all the available properties of that particular entity set to generate the payloads.

After the properties are matched from the properties mentioned in the Exclusion List, they are removed from the List of properties used for generating queries for both the functions.

Also made some changes to the earlier implemented code for Exclusion List in the DirectBuilder class of the entities.py file in the regex_match and apply_restrictions methods for improving the Exclusion_List feature for EntitySets and Properties appearing in the Body.

### CODE REVIEW

File: odfuzz/entities.py [hunk 1]
```diff
@@ -692,9 +690,18 @@ def get_depending_data(self):
 class OrderbyQuery(QueryOption):
     """The search query option."""
 
-    def __init__(self, entity, restrictions):
+    def __init__(self, entity, restrictions, restrictions_group):
```
Comment (phanak-sap) [inline odfuzz/entities.py:693]: Why is here new parameter, restriction_group, while in DirectBuilder the same data are provided just as substructure of restrictions? I would expected that nothing would change here as well.

File: tests/integration/url_generator_only/test_url_generator_only.py [hunk 2]
```diff
@@ -156,38 +168,137 @@ def test_function_imports():
...
+    ("MERGE", "Suppliers(SupplierID=1564408389)?sap-client=500"),
+    ("PUT", "Order_Details_Extendeds(OrderID=-717893126,ProductID=349786362,ProductName='8%C2%A3T%E2%80%A0%C2%81%C3%A1C%C3%8F%C3%96%C3%A5%C5%92%C2%A7%29%C3%83T%C3%9A%C5%92m',UnitPrice=16647857647390.73m,Quantity=-1289,Discount=1.4474711818031789e+20f)?sap-client=500"),
+    ("DELETE", "Alphabetical_list_of_products?sap-client=500")
+])
+
+def test_direct_builder_Uri_unrestricted(method_name, URI):
+    random.seed(30)
+    entities , queryable_factory = builder(method_name)
+    methodList = []
+    methodList.clear()
```
Comment (phanak-sap) [inline tests/integration/url_generator_only/test_url_generator_only.py:227]: is this needed? every parametrized test should be new instance with new clear methodList from line 226

File: odfuzz/entities.py (hunk 1 above)
Comment (deev-pal08) [reply odfuzz/entities.py:693]: The first Parameter (restrictions) calls the forbidden_options method of the class RestrictionGroup which does the not provide the Exclusion_List we require to restrict the Properties. This is just a subpart of the RestrictionsGroup.

It provides us with the YAML file restriction Object:
Exclude:
    $filter:
        Categories:
            - Description
            - Picture
        Employees:
            - Photo
            - HireDate
        $F_ALL$:
            - replace
    $orderby:
        Categories:
            - Picture
            - Description
       
What we actually need is the restrictions passed as a json object to the RestrictionGroup Class when we initialize the RestrictionGroup class.

The second Parameter (restrictions_group) calls the excluded_options method which provides us with the ExclusionList Json Object to actually restrict the properties.

This Method provides us with the actual Json Object which we need to restrict the properties:
exclusion_dict = {"$ENTITY_SET$":{".*": {"Properties":["ProductID"],"Nav_Properties":[]}}}

This is what we require to restrict the properties and not the previous object.

File: tests/integration/url_generator_only/test_url_generator_only.py (hunk 2 above)
Comment (deev-pal08) [reply tests/integration/url_generator_only/test_url_generator_only.py:227]: This won't be required, will remove this in the next commit..

File: odfuzz/entities.py [hunk 3]
```diff
@@ -888,6 +909,30 @@ def generate(self, depending_data):
...
+                    proprty_index = 0
+                    while proprty_index < len(required_properties.proprties):
+                        proprty = required_properties.proprties[proprty_index]
+                        if self._regex_match(self.entity_set._name, proprty.name):
+                            required_properties._proprties.remove(proprty)
+                            self._filterable_proprties.remove(proprty)
+                            proprty_index -= 1
+                        proprty_index += 1
+    
+    def _regex_match(self, entity_name, proprty_name):
```
Comment (phanak-sap) [inline odfuzz/entities.py:925]: same.. rename to better self-documenting name

**Review by phanak-sap — APPROVED** [review]

**Pull request merged**